File: testdata/cross-repo/grpc/client-csharp/PromoCodeClient.cs
// C# gRPC client fixture for IDL cross-repo binding tests.
//
// Mirrors what Grpc.Tools emits via protoc-gen-grpc-csharp: a generated nested
// client class named `<Service>.<Service>Client` constructed with a Grpc.Net
// channel; rpcs are invoked as `*Async` methods on the client instance.
//
// The cbm pass_idl_scan should detect:
//   - The `_client` field typed as `PromoCodeService.PromoCodeServiceClient`
//     (suffix `Client`)
//   - Strip the `Client` suffix → service name `PromoCodeService`
//   - Match against the proto-derived `PromoCodeService` Class node
//   - For each `_client.<RpcName>Async(...)` call within `FetchVoucherAsync`,
//     strip the `Async` suffix and emit a `GRPC_CALLS` edge to the Route
//     `__route__grpc__PromoCodeService/<RpcName>`.

namespace Promo.V1.Client;

// Stand-in for Grpc.Tools-generated PromoCodeServiceClient.
public class PromoCodeServiceClient
{
    public PromoCodeServiceClient(Grpc.Net.Client.GrpcChannel channel) { _channel = channel; }
    private readonly Grpc.Net.Client.GrpcChannel _channel;

    public virtual System.Threading.Tasks.Task<GetVoucherResponse> GetVoucherAsync(
        GetVoucherRequest request, Grpc.Core.CallOptions options = default)
        => System.Threading.Tasks.Task.FromResult(new GetVoucherResponse());

    public virtual System.Threading.Tasks.Task<RedeemVoucherResponse> RedeemVoucherAsync(
        RedeemVoucherRequest request, Grpc.Core.CallOptions options = default)
        => System.Threading.Tasks.Task.FromResult(new RedeemVoucherResponse());
}

public class CheckoutFlow
{
    private readonly PromoCodeServiceClient _client;

    public CheckoutFlow(Grpc.Net.Client.GrpcChannel channel)
    {
        _client = new PromoCodeServiceClient(channel);
    }

    public async System.Threading.Tasks.Task<GetVoucherResponse> FetchVoucherAsync(string voucherId)
    {
        var request = new GetVoucherRequest { VoucherId = voucherId };
        return await _client.GetVoucherAsync(request);
    }

    public async System.Threading.Tasks.Task<bool> RedeemAsync(string voucherId, string userId)
    {
        var req = new RedeemVoucherRequest { VoucherId = voucherId, UserId = userId };
        var resp = await _client.RedeemVoucherAsync(req);
        return resp.Success;
    }
}

public class GetVoucherRequest { public string VoucherId { get; set; } = ""; }
public class GetVoucherResponse { public string VoucherId { get; set; } = ""; public int DiscountPercent { get; set; } public bool IsValid { get; set; } }
public class RedeemVoucherRequest { public string VoucherId { get; set; } = ""; public string UserId { get; set; } = ""; }
public class RedeemVoucherResponse { public bool Success { get; set; } public string ErrorMessage { get; set; } = ""; }

File: testdata/cross-repo/grpc/client-java/PromoCodeClient.java
// Java gRPC client fixture for IDL cross-repo binding tests.
//
// Mirrors what protoc-gen-grpc-java emits: a stub factory `PromoCodeServiceGrpc`
// returns a typed `PromoCodeServiceBlockingStub`. RPC methods are invoked
// as lowerCamelCase methods on the stub.
//
// The cbm pass_idl_scan should detect:
//   - The `stub` variable typed as `PromoCodeServiceGrpc.PromoCodeServiceBlockingStub`
//     (suffix `BlockingStub`)
//   - Strip the `BlockingStub` suffix → service name `PromoCodeService`
//   - Match against the proto-derived `PromoCodeService` Class node
//   - For each `stub.<rpcName>(...)` call, capitalize the first character and
//     emit a `GRPC_CALLS` edge to the Route `__route__grpc__PromoCodeService/<RpcName>`.

package com.example.promo.client;

import io.grpc.Channel;
import io.grpc.ManagedChannelBuilder;

// Stand-in for protoc-gen-grpc-java-generated PromoCodeServiceGrpc.
class PromoCodeServiceGrpc {
    public static PromoCodeServiceBlockingStub newBlockingStub(Channel channel) {
        return new PromoCodeServiceBlockingStub();
    }

    public static class PromoCodeServiceBlockingStub {
        public GetVoucherResponse getVoucher(GetVoucherRequest req) { return new GetVoucherResponse(); }
        public RedeemVoucherResponse redeemVoucher(RedeemVoucherRequest req) { return new RedeemVoucherResponse(); }
    }
}

public class PromoCodeClient {
    private final PromoCodeServiceGrpc.PromoCodeServiceBlockingStub stub;

    public PromoCodeClient(String target) {
        Channel channel = ManagedChannelBuilder.forTarget(target).usePlaintext().build();
        this.stub = PromoCodeServiceGrpc.newBlockingStub(channel);
    }

    public GetVoucherResponse fetchVoucher(String voucherId) {
        GetVoucherRequest request = new GetVoucherRequest();
        request.voucherId = voucherId;
        return stub.getVoucher(request);
    }

    public boolean redeem(String voucherId, String userId) {
        RedeemVoucherRequest req = new RedeemVoucherRequest();
        req.voucherId = voucherId;
        req.userId = userId;
        RedeemVoucherResponse resp = stub.redeemVoucher(req);
        return resp.success;
    }
}

class GetVoucherRequest { public String voucherId; }
class GetVoucherResponse {}
class RedeemVoucherRequest { public String voucherId; public String userId; }
class RedeemVoucherResponse { public boolean success; }
